pkg/volume/csi/csi_mounter.go

```diff
@@ -377,33 +391,40 @@ func (c *csiMountMgr) TearDownAt(dir string) error {
 // 1) if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
 // 2) if fstype is provided and pv.AccessMode == ReadWriteOnly and !c.spec.ReadOnly then apply fsgroup
 func (c *csiMountMgr) applyFSGroup(fsType string, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy) error {
-	if fsGroup != nil {
-		if fsType == "" {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, fsType not provided"))
-			return nil
-		}
+	if c.fsGroupPolicy == storage.FileFSGroupPolicy || fsGroup != nil {
+
+		// If the FSGroupPolicy is ReadWriteOnceWithFSTypeFSGroupPolicy perform additional checks
+		// to determine if we should proceed with modifying the fsGroup.
+		if c.fsGroupPolicy == storage.ReadWriteOnceWithFSTypeFSGroupPolicy {
+			if fsType == "" {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, fsType not provided"))
+				return nil
+			}
 
-		accessModes := c.spec.PersistentVolume.Spec.AccessModes
-		if c.spec.PersistentVolume.Spec.AccessModes == nil {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, access modes not provided"))
-			return nil
-		}
-		if !hasReadWriteOnce(accessModes) {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, only support ReadWriteOnce access mode"))
-			return nil
-		}
+			accessModes := c.spec.PersistentVolume.Spec.AccessModes
+			if c.spec.PersistentVolume.Spec.AccessModes == nil {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, access modes not provided"))
+				return nil
+			}
+			if !hasReadWriteOnce(accessModes) {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, only support ReadWriteOnce access mode"))
+				return nil
+			}
 
-		if c.readOnly {
-			klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, volume is readOnly"))
-			return nil
+			if c.readOnly {
+				klog.V(4).Info(log("mounter.SetupAt WARNING: skipping fsGroup, volume is readOnly"))
+				return nil
+			}
 		}
 
 		err := volume.SetVolumeOwnership(c, fsGroup, fsGroupChangePolicy)
```

Comment:
I think this will cause `SetVolumeOwnership` to be called for `readonly` volumes.